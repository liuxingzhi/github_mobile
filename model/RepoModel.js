const emptyRepoData = {
    name: '',
    url: '',
    description: '',
    updatedAt: '',
    primaryLanguage: {
        name: '',
        color: '#FFFFFF', // White, invisible by default
    },
    stargazers: {
        totalCount: 0,
    },
    forkCount: 0
}

class RepoModel {
    constructor(repoData = emptyRepoData) {
        if (repoData == null){
            repoData = emptyRepoData
        }
        this.name = repoData.name != null ? repoData.name.trim() : ''
        this.url = repoData.url != null ? repoData.url : ''
        this.description = repoData.description != null ? repoData.description : ''
        this.updatedAt = repoData.updatedAt != null ? repoData.updatedAt : ''
        this.primaryLanguage = repoData.primaryLanguage != null ? repoData.primaryLanguage.name : ''
        this.primaryLanguageColor = repoData.primaryLanguage != null ? repoData.primaryLanguage.color : '#ffffff'
        this.starCount = repoData.stargazers != null ? repoData.stargazers.totalCount : 0
        this.forkCount = repoData.forkCount != null ? repoData.forkCount : 0
    }

    searchKeyCaseInsensitive(key){
        return this.name.toLowerCase().includes(key.toLowerCase());
    }
}

export {RepoModel, emptyRepoData}