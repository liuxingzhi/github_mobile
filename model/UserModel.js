const emptyUserData = {
    avatarUrl: '',
    name: '',
    login: '',
    bio: '',
    websiteUrl: '',
    email: '',
    repositories: {totalCount: ''},
    followers: {totalCount: ''},
    following: {totalCount: ''},
    createDate: '',
}

class UserModel {
    constructor(userData = emptyUserData) {
        if (userData == null){
            userData = emptyUserData;
        }
        this.profileImage = userData.avatarUrl != null ? userData.avatarUrl : ''
        this.name = userData.name != null ? userData.name : ''
        this.username = userData.login != null ? userData.login : ''
        this.bio = userData.bio
        this.website = userData.websiteUrl != null ? userData.websiteUrl : ''
        this.email = userData.email != null ? userData.email : ''
        this.publicReposCount = userData.repositories != null ? userData.repositories.totalCount : 0
        this.followersCount = userData.followers != null ? userData.followers.totalCount : 0
        this.followingCount = userData.following != null ? userData.following.totalCount : 0
        this.createDate = userData.createdAt
    }

    set createDate(createdAt) {
        this._createDate = createdAt != null ? createdAt : ''
    }

    get createDate() {
        return this._createDate.split('T')[0]
    }

    set bio(userBio) {
        this._bio = userBio != null ? userBio : ''
    }

    get bio() {
        return this._bio.trim()
    }

    searchKeyCaseInsensitive(key){
        return this.username.toLowerCase().includes(key.toLowerCase());
    }
}

export {UserModel, emptyUserData}