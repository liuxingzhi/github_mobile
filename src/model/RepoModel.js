/**
 * default empty data structure for repo
 * @type {{stargazers: {totalCount: number}, name: string, description: string, forkCount: number, url: string, primaryLanguage: {color: string, name: string}, updatedAt: string}}
 */
const emptyRepoData = {
    name: '',
    url: '',
    description: '',
    updatedAt: '',
    primaryLanguage: {
        name: '',
        color: '#FFFFFF', // White, invisible by default
    },
    stargazers: {
        totalCount: 0,
    },
    forkCount: 0
}

/**
 * Class model for repository, used by repo screen
 */
class RepoModel {
    constructor(repoData = emptyRepoData) {
        if (repoData == null){
            repoData = emptyRepoData
        }
        this.name = repoData.name != null ? repoData.name.trim() : ''
        this.url = repoData.url != null ? repoData.url : ''
        this.description = repoData.description != null ? repoData.description : ''
        this.updatedAt = repoData.updatedAt != null ? repoData.updatedAt : ''
        this.primaryLanguage = repoData.primaryLanguage != null ? repoData.primaryLanguage.name : ''
        this.primaryLanguageColor = repoData.primaryLanguage != null ? repoData.primaryLanguage.color : '#ffffff'
        this.starCount = repoData.stargazers != null ? repoData.stargazers.totalCount : 0
        this.forkCount = repoData.forkCount != null ? repoData.forkCount : 0
        this.owner = repoData.owner != null ? repoData.owner.login : ''
    }

    searchKeyCaseInsensitive(key){
        return this.name.toLowerCase().includes(key.toLowerCase());
    }
}

export {RepoModel, emptyRepoData}