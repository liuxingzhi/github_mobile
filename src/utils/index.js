// export { loadFollowerData, loadFollowingData, loadRepoData, loadUserData } from "./service.js";
// export { githubQuery, parseNodesToModel } from "./util.js";

export * from "./service.js";
export * from "./util.js";