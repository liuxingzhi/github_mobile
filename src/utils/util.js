import * as ENV from "../env.js"
import fetch from 'cross-fetch'
import {emptyUserData, UserModel} from "../model/UserModel.js"

/**
 * A wrap to the github API call
 * @param query
 * @param token
 * @returns {Promise<Response|{}>}
 */
async function githubQuery(query, token) {
    let url = 'https://api.github.com/graphql'
    return await fetch(url, {
        method: 'POST',
        headers: {
            Authorization: `bearer ${ENV.USER_AUTH_TOKEN}`,
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            query: query,
        }),
    }).then(response => {
        if (response.ok) {
            return response.json();
        } else {
            switch (response.status){
                case 401:
                    console.log(new Error(`401 Unauthorized`))
                case 403:
                    console.log (new Error(`403 Forbidden`))
                default:
                    return response.json()
            }
        }
    }).catch(error => {
        console.log(error)
        return {}
    })
}

/**
 * A helper function to convert nodes array to corresponding list of model
 * @param nodesArray
 * @param Model: A class instance, the returned model list will be of the class type passed in
 * @returns {[]}
 */
function parseNodesToModel(nodesArray, Model) {
    let modelList = []
    for (const node of nodesArray) {
        modelList.push(new Model(node))
    }
    return modelList
}

export {githubQuery, parseNodesToModel}
