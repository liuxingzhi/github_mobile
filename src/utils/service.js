/**
 * This file contain 4 different kinds of services,
 * loadUserData
 * loadFollowerData
 * loadFollowingData
 * loadRepoData
 */
import {githubQuery, parseNodesToModel} from "./util.js";
import {emptyUserData, UserModel} from "../model/UserModel.js";
import {emptyRepoData, RepoModel} from "../model/RepoModel.js";


async function loadUserData(username) {
    let query = `query { 
                user(login:"${username}") { 
                    login
                    bio
                    avatarUrl
                    websiteUrl
                    createdAt
                    name
                    email
                    repositories(first: 100) {
                        totalCount
                    }
                    followers {
                        totalCount
                    }
                    following {
                        totalCount
                    }
                }
        }`

    try {
        // load data
        let response = await githubQuery(query);
        let userData = response.data.user
        return new UserModel(userData)
    }catch (error){
        // use empty data when error
        // console.log(error)
        return new UserModel(emptyUserData)
    }
}

async function loadRepoData(username) {
    let query = `query {
                        user(login:"${username}") {
                            repositories(first: 100, orderBy: {field: UPDATED_AT, direction: DESC}){
                                nodes{
                                    owner{
                                      login
                                    }
                                    name
                                    url
                                    description
                                    updatedAt
                                    primaryLanguage{
                                        name
                                        color
                                    }
                                    stargazers{
                                        totalCount
                                    }
                                    forkCount
                                }
                            }
                        }
                    }`
    try {
        // load data
        let response = await githubQuery(query);
        let repoDataArray = response.data.user.repositories.nodes
        let repos = parseNodesToModel(repoDataArray, RepoModel)
        return repos
    }catch (error){
        // use empty data when error
        // console.log(error)
        return [new RepoModel(emptyRepoData)]
    }
}

async function loadFollowerData(username) {
    let query = `
        {
          user(login: "${username}") {
            login
            bio
            followers(first: 100) {
              nodes {
                login
                bio
                name
                avatarUrl
                repositories {
                  totalCount
                }
              }
            }
          }
        }`

    try {
        // load data
        let response = await githubQuery(query);
        let userDataArray = response.data.user.followers.nodes
        let userList = parseNodesToModel(userDataArray, UserModel)
        userList.sort(function (a, b) {
            return a.publicReposCount - b.publicReposCount
        })
        return userList
    }catch (error){
        // use empty data when error
        // console.log(error)
        return [new UserModel(emptyUserData)]
    }
}

async function loadFollowingData(username) {
    let query = `
        {
          user(login: "${username}") {
            login
            bio
            following(first: 100) {
              nodes {
                login
                name
                bio
                avatarUrl
                repositories {
                  totalCount
                }
              }
            }
          }
        }`

    try {
        // load data
        let response = await githubQuery(query);
        let userDataArray = response.data.user.following.nodes
        let userList = parseNodesToModel(userDataArray, UserModel)
        userList.sort(function (a, b) {
            return a.publicReposCount - b.publicReposCount
        })
        return userList
    }catch (error){
        // use empty data when error
        // console.log(error)
        return [new UserModel(emptyUserData)]
    }
}

export {loadFollowerData, loadFollowingData, loadRepoData, loadUserData}