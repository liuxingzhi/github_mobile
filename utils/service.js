/**
 * This file contain 4 different kinds of services,
 * loadUserData
 * loadFollowerData
 * loadFollowingData
 * loadRepoData
 */
import {githubQuery, parseNodesToModel} from "./util";
import {emptyUserData, UserModel} from "../model/UserModel";
import {emptyRepoData, RepoModel} from "../model/RepoModel";

function loadUserData(username) {
    let query = `query { 
                user(login:"${username}") { 
                    login
                    bio
                    avatarUrl
                    websiteUrl
                    createdAt
                    name
                    email
                    repositories(first: 100) {
                        totalCount
                    }
                    followers {
                        totalCount
                    }
                    following {
                        totalCount
                    }
                }
        }`

    githubQuery(query)
        .then(response => {
            let userData = response.data.user
            return new UserModel(userData)
        })
        .catch(error => {
            console.log(error)
            return new UserModel(emptyUserData)
        })
}

function loadRepoData(username) {
    let query = `query {
                        user(login:"${username}") {
                            repositories(first: 100, orderBy: {field: UPDATED_AT, direction: DESC}){
                                nodes{
                                    name
                                    url
                                    description
                                    updatedAt
                                    primaryLanguage{
                                        name
                                        color
                                    }
                                    stargazers{
                                        totalCount
                                    }
                                    forkCount
                                }
                            }
                        }
                    }`
    githubQuery(query)
        .then(response => {
            let repoDataArray = response.data.user.repositories.nodes
            let repos = parseNodesToModel(repoDataArray, RepoModel)
            return repos
        })
        .catch(error => {
            console.log(error)
            return [new RepoModel(emptyRepoData)]
        });
}

function loadFollowerData(username) {
    let query = `
        {
          user(login: "${username}") {
            login
            bio
            followers(first: 100) {
              nodes {
                login
                bio
                avatarUrl
                repositories {
                  totalCount
                }
              }
            }
          }
        }`

    githubQuery(query)
        .then(response => {
            // console.log(response.data.user.followers.nodes)
            let userDataArray = response.data.user.followers.nodes
            let userList = parseNodesToModel(userDataArray, UserModel)
            userList.sort(function (a, b) {
                return a.publicReposCount - b.publicReposCount
            })
            return userList
        })
        .catch(error => {
            console.log(error)
            return [new UserModel(emptyUserData)]
        })
}

function loadFollowingData() {
    let query = `
        {
          user(login: "${this.currentUser}") {
            login
            bio
            following(first: 100) {
              nodes {
                login
                bio
                avatarUrl
                repositories {
                  totalCount
                }
              }
            }
          }
        }`

    githubQuery(query)
        .then(response => {
            // console.log(response.data.user.followers.nodes)
            let userDataArray = response.data.user.following.nodes
            let userList = parseNodesToModel(userDataArray, UserModel)
            userList.sort(function (a, b) {
                return a.publicReposCount - b.publicReposCount
            })
            return userList
        })
        .catch(error => {
            console.log(error)
            return [new UserModel(emptyUserData)]
        })
}

export {loadFollowerData, loadFollowingData, loadRepoData, loadUserData}