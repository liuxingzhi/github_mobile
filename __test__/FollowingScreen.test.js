import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, {shallow, render, mount} from 'enzyme';
import toJson from "enzyme-to-json";
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({adapter: new Adapter()});

import {githubQuery} from '../src/utils/util';
import ProfileScreen from "../src/screens/Profile";
import RepositoryScreen from "../src/screens/Repo";
import FollowerScreen from "../src/screens/Following";
import FollowingScreen from "../src/screens/Follower";
import {RepoModel} from "../src/model/RepoModel";
import {UserModel} from "../src/model/UserModel";
import ShallowWrapper from "enzyme/src/ShallowWrapper";

const fakeFollowingResponse = JSON.stringify({
    "data":
        {
            "user":
                {
                    "login":
                        "liuxingzhi",
                    "bio":
                        "CS student in UIUC, passionate programmer and Go player.",
                    "following": {
                        "nodes": [
                            {
                                "login": "yishn",
                                "bio": "(he/him) Software developer, mathematician at heart, wannabe artist, and Go player.",
                                "avatarUrl": "https://avatars2.githubusercontent.com/u/9217349?u=8f87121c9ff64d666a0c43ae34e5de6d36bd0955&v=4",
                                "repositories": {
                                    "totalCount": 48
                                }
                            },
                            {
                                "login": "minweny",
                                "bio": "",
                                "avatarUrl": "https://avatars1.githubusercontent.com/u/25225921?u=ca6e2f79a36f731a5f45853ce1f0425c26005b5f&v=4",
                                "repositories": {
                                    "totalCount": 24
                                }
                            },
                            {
                                "login": "CryoliteZ",
                                "bio": "Open source, Open mind\r\n",
                                "avatarUrl": "https://avatars0.githubusercontent.com/u/13447379?u=381b25ba3481853f08ea1160ed8f285e41b28c96&v=4",
                                "repositories": {
                                    "totalCount": 48
                                }
                            },
                            {
                                "login": "buchijiang",
                                "bio": "",
                                "avatarUrl": "https://avatars1.githubusercontent.com/u/13143897?u=262883b491ebf9566a3427e64dd341168f8ab7de&v=4",
                                "repositories": {
                                    "totalCount": 11
                                }
                            },
                            {
                                "login": "BEWINDOWEB",
                                "bio": "Machine Learning 10% +\r\nJava Web 10% +\r\nDistributed System 80%",
                                "avatarUrl": "https://avatars1.githubusercontent.com/u/6178098?u=a3bcc08dee33b91da3b4e5ef4b75763a48223e83&v=4",
                                "repositories": {
                                    "totalCount": 12
                                }
                            },
                            {
                                "login": "LukeLR",
                                "bio": "",
                                "avatarUrl": "https://avatars0.githubusercontent.com/u/3177243?u=aa20bae8d9ad8f1416a3bdbd452b7efa5da171f9&v=4",
                                "repositories": {
                                    "totalCount": 46
                                }
                            },
                        ]
                    }
                }
        }
})

test('test following renders correctly', async () => {
    fetch.mockResponse(() =>  Promise.resolve(fakeFollowingResponse))
    const wrapper = shallow(<FollowingScreen navigation={{addListener: () => null}}
                                             setCurrentUserHandler={(...args) => null}
                                             getCurrentUser={(...args) => null}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
})

test('test following render when api is down', async () => {
    fetch.mockResponse(() => Promise.reject("API is down"))
    const wrapper = shallow(<FollowingScreen navigation={{addListener: () => null}}
                                             setCurrentUserHandler={(...args) => null}
                                             getCurrentUser={(...args) => null}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
})

test('test following render when api returns bad format data', async () => {
    fetch.mockResponse(() => Promise.resolve({}))
    const wrapper = shallow(<FollowingScreen navigation={{addListener: () => null}}
                                             setCurrentUserHandler={(...args) => null}
                                             getCurrentUser={(...args) => null}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
})