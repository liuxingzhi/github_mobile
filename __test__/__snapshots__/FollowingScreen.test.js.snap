// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test following render when api is down 1`] = `
<Styled(Container)>
  <LoadingTheme
    visibility={true}
  />
  <CustomerHeader
    navigation={
      Object {
        "addListener": [Function],
      }
    }
    setFilterCriteriaHandler={[Function]}
    title="Follower"
  />
  <Styled(Content)>
    <Styled(List)>
      <Styled(ListItem)
        avatar={true}
        key="0"
      >
        <Styled(Left)>
          <ForwardRef
            onPress={[Function]}
          >
            <Styled(Thumbnail)
              source={
                Object {
                  "uri": "",
                }
              }
            />
          </ForwardRef>
        </Styled(Left)>
        <Styled(Body)>
          <ForwardRef
            onPress={[Function]}
          >
            <Styled(Text) />
          </ForwardRef>
          <Styled(Text)
            note={true}
          />
        </Styled(Body)>
      </Styled(ListItem)>
    </Styled(List)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`test following render when api returns bad format data 1`] = `
<Styled(Container)>
  <LoadingTheme
    visibility={true}
  />
  <CustomerHeader
    navigation={
      Object {
        "addListener": [Function],
      }
    }
    setFilterCriteriaHandler={[Function]}
    title="Follower"
  />
  <Styled(Content)>
    <Styled(List)>
      <Styled(ListItem)
        avatar={true}
        key="0"
      >
        <Styled(Left)>
          <ForwardRef
            onPress={[Function]}
          >
            <Styled(Thumbnail)
              source={
                Object {
                  "uri": "",
                }
              }
            />
          </ForwardRef>
        </Styled(Left)>
        <Styled(Body)>
          <ForwardRef
            onPress={[Function]}
          >
            <Styled(Text) />
          </ForwardRef>
          <Styled(Text)
            note={true}
          />
        </Styled(Body)>
      </Styled(ListItem)>
    </Styled(List)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`test following renders correctly 1`] = `
<Styled(Container)>
  <LoadingTheme
    visibility={true}
  />
  <CustomerHeader
    navigation={
      Object {
        "addListener": [Function],
      }
    }
    setFilterCriteriaHandler={[Function]}
    title="Follower"
  />
  <Styled(Content)>
    <Styled(List)>
      <Styled(ListItem)
        avatar={true}
        key="0"
      >
        <Styled(Left)>
          <ForwardRef
            onPress={[Function]}
          >
            <Styled(Thumbnail)
              source={
                Object {
                  "uri": "",
                }
              }
            />
          </ForwardRef>
        </Styled(Left)>
        <Styled(Body)>
          <ForwardRef
            onPress={[Function]}
          >
            <Styled(Text) />
          </ForwardRef>
          <Styled(Text)
            note={true}
          />
        </Styled(Body)>
      </Styled(ListItem)>
    </Styled(List)>
  </Styled(Content)>
</Styled(Container)>
`;
